keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,517 @@
+container.
+
+Instead, k8s can provide a new mechanism that will be used to instantiate
+env vars from the specified file. The user can generate an env file using an
+`initContainer`. The container can then use the `fileRef` field to reference
+the file from which it should initialize the env vars. Behind the scene, kubelet will
+parse the env file and instantiate these env vars during container creation. It
+does not mount this file onto the main container. In the example below,
+the Pod's output includes `CONFIG_VAR=HELLO`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo CONFIG_VAR="HELLO" > /etc/config/config.env']
+    volumeMounts:
+    - name: data
+      mountPath: /etc/config
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    envFrom:
+    - fileRef:
+        path: config.env
+        volumeName: data
+  restartPolicy: Never
+  volumes:
+  - name: data
+    emptyDir: {}
+```
+
+#### Story 2
+
+Each k8s Node is instantiated with an env file that contains information about
+the Node (for eg, instance type, failure zones). The user would like to
+provide some of the information in this file to the vendor's container via
+env vars. For example, consider that each node has the following env file:
+
+```
+CONFIG_VAR=HELLO
+CONFIG_VAR_A=WORLD
+...
+```
+
+The user can use the `fileKeyRef` field to select a subset of these env vars
+that it wants to provide to the app. Behind the scene, kubelet will parse the
+env file on the host and instantiate these env vars when creating the main
+container. It does not mount this hostPath volume onto the main container.
+In this case, the Pod's output includes `CONFIG_VAR=HELLO`
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: node.env
+          volumeName: data
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: data
+    hostPath: /etc/config
+```
+
+## Design Details
+
+Pods and PodTemplate will include `FileEnvSource` field under [`EnvFromSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvFromSource)
+and `FileKeySelector` field under [`EnvVarSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvVarSource)
+that can be set for an individual container.
+
+### Pod API
+
+`FileEnvSource` field will allow the users to populate environment variables
+from the specified file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+```
+type EnvFromSource struct {
+    ...
+    // The file to select from
```

Comment:
This needs to be a full specification.  At what time is the file read?  What happens if the file is changed after this is read?  If the container restarts, will the file be re-read?  How big can the file be?  What user-ID are we using to read the file (the container's effective UID might not be allowed to read it!)

Reply:
Done.